File: addon.cc
#include <nan.h>
#include <iostream>
#include <Eigen/Dense>
#include <Eigen/Eigenvalues> 
#include <complex>
using namespace v8;


NAN_METHOD(MatMul) {
	NanScope();
	using CMf = Eigen::Map <const Eigen::MatrixXf >;
	using Mf = Eigen::Map <Eigen::MatrixXf >;
	//Map<MatrixXf> mf(pf, rows, columns);

	if (args.Length() < 7) {
		NanThrowTypeError("Wrong number of arguments");
		NanReturnUndefined();
	}

	if (!args[0]->IsUint32() || !args[1]->IsUint32() || !args[2]->IsFloat32Array() ||
		!args[3]->IsUint32() || !args[4]->IsUint32() || !args[5]->IsFloat32Array() ||
		!args[6]->IsFloat32Array()) {
		NanThrowTypeError("Wrong arguments");
		NanReturnUndefined();
	}
	size_t rows1(args[0]->Uint32Value());
	size_t cols1(args[1]->Uint32Value());
	Local<Float32Array> arr1 = args[2].As<Float32Array>();
	Local<ArrayBuffer> buffer1 = arr1->Buffer();
	ArrayBuffer::Contents contents1 = buffer1->Externalize();
	float* data1 = static_cast<float*>(contents1.Data());
	std::cout << "After cast." << std::endl;
	CMf first(data1, rows1, cols1);

	size_t rows2(args[3]->Uint32Value());
	size_t cols2(args[4]->Uint32Value());
	Local<Float32Array> arr2 = args[5].As<Float32Array>();
	Local<ArrayBuffer> buffer2 = arr2->Buffer();
	ArrayBuffer::Contents contents2 = buffer2->Externalize();
	float* data2 = static_cast<float*>(contents2.Data());
	CMf second(data2, rows2, cols2);

	Local<Float32Array> resFArray = args[6].As<Float32Array>();
	Local<ArrayBuffer> resBuff = resFArray->Buffer();
	ArrayBuffer::Contents resContents = resBuff->Externalize();
	float* resRawData = static_cast<float*>(resContents.Data());
	
	Mf res(resRawData, rows1, cols2);
	res = second * first;
	//ArrayBuffer::Contents c
	Local<Float32Array> num = NanNew(resFArray);
	std::cout << "Before return." << std::endl;
	NanReturnValue(num);
}

NAN_METHOD(MatMul2){
	NanScope();
	using CMf = Eigen::Map <const Eigen::MatrixXf >;
	using Mf = Eigen::Map <Eigen::MatrixXf >;
	//Map<MatrixXf> mf(pf, rows, columns);

	if (args.Length() < 7) {
		NanThrowTypeError("Wrong number of arguments");
		NanReturnUndefined();
	}

	if (!args[0]->IsUint32() || !args[1]->IsUint32() ||/* !args[2]->IsFloat32Array() ||*/
		!args[3]->IsUint32() || !args[4]->IsUint32()   /*||  !args[5]->IsFloat32Array() ||
													   !args[6]->IsFloat32Array()*/) {
		NanThrowTypeError("Wrong arguments");
		NanReturnUndefined();
	}
	size_t rows1(args[0]->Uint32Value());
	size_t cols1(args[1]->Uint32Value());
	
	
	Local<Object> matrix1Buff = args[2].As<Object>();
	float *data1 = nullptr;
	size_t length1 = 0;

	if (matrix1Buff->HasIndexedPropertiesInExternalArrayData()) {
		length1 = size_t(matrix1Buff->GetIndexedPropertiesExternalArrayDataLength());
		data1 = static_cast<float*>(matrix1Buff->GetIndexedPropertiesExternalArrayData());
	}
	CMf first(data1, rows1, cols1);

	size_t rows2(args[3]->Uint32Value());
	size_t cols2(args[4]->Uint32Value());

	Local<Object> matrix2Buff = args[5].As<Object>();
	float *data2 = nullptr;
	size_t length2 = 0;

	if (matrix2Buff->HasIndexedPropertiesInExternalArrayData()) {
		length2 = size_t(matrix2Buff->GetIndexedPropertiesExternalArrayDataLength());
		data2 = static_cast<float*>(matrix2Buff->GetIndexedPropertiesExternalArrayData());
	}

	CMf second(data2, rows2, cols2);
	
	Local<Object> matrixResBuff = args[6].As<Object>();
	float *resRawData = nullptr;
	size_t lengthRes = 0;

	if (matrixResBuff->HasIndexedPropertiesInExternalArrayData()) {
		lengthRes = size_t(matrixResBuff->GetIndexedPropertiesExternalArrayDataLength());
		resRawData = static_cast<float*>(matrixResBuff->GetIndexedPropertiesExternalArrayData());
	}
	
	Mf res(resRawData, rows1, cols2);
	res = second * first;
	Local<Boolean> b = NanNew(true);
	NanReturnValue(b);
}


NAN_METHOD(GetEigenValues){
	using CMf = Eigen::Map <const Eigen::MatrixXf >;
	using Mf = Eigen::Map <Eigen::MatrixXf >;
	using EMf = Eigen::Map <Eigen::EigenSolver<const Eigen::MatrixXf>::EigenvalueType>;

	if (args.Length() != 4) {
		NanThrowTypeError("Wrong number of arguments");
		NanReturnUndefined();
	}

	if (!args[0]->IsUint32() || !args[1]->IsUint32()) {
		NanThrowTypeError("Wrong arguments");
		NanReturnUndefined();
	}
	size_t rows1(args[0]->Uint32Value());
	size_t cols1(args[1]->Uint32Value());

	Local<Object> matrix1Buff = args[2].As<Object>();
	float *data1 = nullptr;
	size_t length1 = 0;

	if (matrix1Buff->HasIndexedPropertiesInExternalArrayData()) {
		length1 = size_t(matrix1Buff->GetIndexedPropertiesExternalArrayDataLength());
		data1 = static_cast<float*>(matrix1Buff->GetIndexedPropertiesExternalArrayData());
	}
	CMf first(data1, rows1, cols1);

	Local<Object> matrixResBuff = args[3].As<Object>();
	float *resRawData = nullptr;
	size_t lengthRes = 0;

	if (matrixResBuff->HasIndexedPropertiesInExternalArrayData()) {
		lengthRes = size_t(matrixResBuff->GetIndexedPropertiesExternalArrayDataLength());
		resRawData = static_cast<float*>(matrixResBuff->GetIndexedPropertiesExternalArrayData());
	}
	EMf eigenResults(reinterpret_cast<std::complex<float>*>(resRawData)/*very ugly but well defined afaik */, rows1, 1);
	Eigen::EigenSolver<Eigen::MatrixXf> eigenSolver(first, false);

	eigenResults = eigenSolver.eigenvalues();
	Local<Boolean> b = NanNew(true);
	NanReturnValue(b);
}


void Init(Handle<Object> exports) {
	exports->Set(NanNew("MatMul"), NanNew<FunctionTemplate>(MatMul)->GetFunction());
	exports->Set(NanNew("MatMul2"), NanNew<FunctionTemplate>(MatMul2)->GetFunction());
	exports->Set(NanNew("GetEigenValues"), NanNew<FunctionTemplate>(GetEigenValues)->GetFunction());
}

NODE_MODULE(addon, Init)

File: addon.js
var addon = require('bindings')('addon.node')

var mat1 = new Float32Array([1.1,1.1,1.1,1.1]);
var mat2 = new Float32Array([1.0,1.0,1.0,1.0]);
var res = new Float32Array([0.0,0.0,0.0,0.0]);

console.log('Before ', mat1, mat2, res)
console.log('Res:', addon.MatMul(2,2,mat1,2,2,mat2, res))
console.log('After', mat1, mat2, res)

console.log('---------------------Before smalloc--------------------------')
var smalloc = require('smalloc');
var data1 = smalloc.alloc(4, smalloc.Types.Float);
var data2 = smalloc.alloc(4, smalloc.Types.Float);
var data3 = smalloc.alloc(4, smalloc.Types.Float);

data1[0] = 1.0;
data1[1] = 1.0;
data1[2] = 1.0;
data1[3] = 1.0;

console.log("-------------data1--------------------")
console.log(data1[0], " ",data1[1]);
console.log(data1[2], " ",data1[3]);

data2[0] = 1.0;
data2[1] = 1.0;
data2[2] = 1.0;
data2[3] = 1.0;

console.log("-------------data2--------------------")
console.log(data2[0], " ",data2[2]);
console.log(data2[1], " ",data2[3]);

for (var i = 0; i < 9; i++) {
	data3[i] = 0.0;
}

console.log("-------------data3--------------------")
console.log(data3[0], " ",data3[2]);
console.log(data3[1], " ",data3[3]);

console.log('Res:', addon.MatMul2(2,2,data1,2,2,data2, data3));
console.log('After');

console.log("-------------data1--------------------")
console.log(data1[0], " ",data1[2]);
console.log(data1[1], " ",data1[3]);
console.log("-------------data2--------------------")
console.log(data2[0], " ",data2[2]);
console.log(data2[1], " ",data2[3]);
console.log("-------------data3--------------------")
console.log(data3[0], " ",data3[2]);
console.log(data3[1], " ",data3[3]);

console.log("\n\n-------------testing eigenvalues--------------------\n")
var eigenTest = smalloc.alloc(9, smalloc.Types.Float);
var eigenTestResult = smalloc.alloc(6, smalloc.Types.Float);
for (var i = 0; i < 9; i++) {
	eigenTest[i] = i + 1.0;
}
eigenTest[0] = 0.8;
eigenTest[1] = 0.05;
eigenTest[2] = 0.0;
eigenTest[3] = -0.05;
eigenTest[4] = 1.0;
eigenTest[5] = 0.05;
eigenTest[6] = -0.05;
eigenTest[7] = 0.0;
eigenTest[8] = 1.0;

addon.GetEigenValues(3,3,eigenTest,eigenTestResult);

for (var i = 0; i < 5 ; i+=2) {
	console.log(eigenTestResult[i] +"+i " + eigenTestResult[i+1]);
}





